File: server/src/main/com/broll/gainea/server/core/battle/Battle.java
package com.broll.gainea.server.core.battle;

import com.broll.gainea.misc.RandomUtils;
import com.broll.gainea.server.core.objects.Unit;

import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

public class Battle {

    private RollResult attackerRolls;
    private RollResult defenderRolls;

    private List<Unit> attackers;
    private List<Unit> defenders;

    public Battle(List<Unit> attackingUnits, List<Unit> defendingUnits, RollResult attackerRolls, RollResult defenderRolls) {
        this.attackers = attackingUnits;
        this.defenders = defendingUnits;
        this.attackerRolls = attackerRolls;
        this.defenderRolls = defenderRolls;
    }

    public Battle(BattleContext context, List<Unit> attackingUnits, List<Unit> defendingUnits) {
        this(attackingUnits, defendingUnits, new RollResult(context, attackingUnits),
                new RollResult(context, defendingUnits));
    }

    public FightResult fight() {
        int attacks = attackerRolls.count();
        int blocks = defenderRolls.count();
        int battleSize = Math.min(attacks, blocks);
        List<RollResult.Roll> allAttackRolls = attackerRolls.getRolls();
        List<RollResult.Roll> allBlockRolls = defenderRolls.getRolls();
        List<RollResult.Roll> attackRolls = allAttackRolls.stream().limit(battleSize).collect(Collectors.toList());
        List<RollResult.Roll> blockRolls = allBlockRolls.stream().limit(battleSize).collect(Collectors.toList());
        FightResult result = new FightResult(allAttackRolls.stream().map(it -> it.value).collect(Collectors.toList()),
                allBlockRolls.stream().map(it -> it.value).collect(Collectors.toList()));
        for (int i = 0; i < battleSize; i++) {
            RollResult.Roll attack = attackRolls.get(i);
            RollResult.Roll block = blockRolls.get(i);
            if (attack.value > block.value) {
                dealDamage(result, attack, attackers, defenders);
            } else {
                dealDamage(result, block, defenders, attackers);
            }
        }
        List<Unit> deadAttackers = attackers.stream().filter(Unit::isDead).collect(Collectors.toList());
        List<Unit> deadDefenders = defenders.stream().filter(Unit::isDead).collect(Collectors.toList());
        result.killed(deadAttackers, deadDefenders);
        return result;
    }

    private void dealDamage(FightResult result, RollResult.Roll winningRoll, List<Unit> sourceUnits, List<Unit> targetUnits) {
        Unit source = winningRoll.source;
        if (source == null) {
            source = RandomUtils.pickRandom(sourceUnits);
        }
        Unit target = getDamageTarget(targetUnits);
        if (target != null) {
            damage(result, source, target);
        }
    }

    protected void damage(FightResult result, Unit source, Unit target) {
        boolean lethal = target.takeDamage();
        if (lethal) {
            source.addKill();
        }
        result.damage(source, target, lethal);
    }


    private Unit getDamageTarget(List<Unit> targetUnits) {
        List<Unit> targets = targetUnits.stream().filter(Unit::isAlive).collect(Collectors.toList());
        //shuffe for damage (so that same powerlevel units get hit randomly)
        Collections.shuffle(targets);
        //sort ascending (so that weakest power level units die first)
        return targets.stream().min(Comparator.comparingInt(o -> o.getPower().getValue())).orElse(null);
    }

}

File: server/src/main/com/broll/gainea/server/core/battle/Battle.kt
package com.broll.gainea.server.core.battle

import com.broll.gainea.misc.RandomUtils
import com.broll.gainea.server.core.battle.RollResult.Roll
import com.broll.gainea.server.core.objects.Unit

open class Battle(
    private val context: BattleContext,
    private val attackerRolls: RollResult,
    private val defenderRolls: RollResult,
    private val attackers: List<Unit> = context.aliveAttackers,
    private val defenders: List<Unit> = context.aliveDefenders,
) {
    constructor(context: BattleContext) :
            this(
                context = context,
                attackerRolls = RollResult(context, context.aliveAttackers),
                defenderRolls = RollResult(context, context.aliveDefenders)
            )

    fun fight(): FightResult {
        val attacks = attackerRolls.count()
        val blocks = defenderRolls.count()
        val battleSize = Math.min(attacks, blocks)
        val allAttackRolls = attackerRolls.rolls
        val allBlockRolls = defenderRolls.rolls
        val attackRolls = allAttackRolls.take(battleSize)
        val blockRolls = allBlockRolls.take(battleSize)
        val result = FightResult(allAttackRolls, allBlockRolls)
        for (i in 0 until battleSize) {
            val attack = attackRolls[i]
            val block = blockRolls[i]
            if (attack.value > block.value) {
                dealDamage(result, attack, attackers, defenders)
            } else {
                dealDamage(result, block, defenders, attackers)
            }
        }
        val deadAttackers = attackers.filter { it.dead }
        val deadDefenders = defenders.filter { it.dead }
        result.killed(deadAttackers, deadDefenders)
        return result
    }

    private fun dealDamage(
        result: FightResult,
        winningRoll: Roll,
        sourceUnits: List<Unit>,
        targetUnits: List<Unit>
    ) {
        val target = getDamageTarget(targetUnits)
        target?.let {
            damage(result, winningRoll.source ?: RandomUtils.pickRandom(sourceUnits), it)
        }
    }

    protected open fun damage(result: FightResult, source: Unit, target: Unit) {
        val lethal = target.takeDamage()
        if (lethal) {
            source.killedEnemy(target)
        }
        result.damage(source, target, lethal)
    }

    private fun getDamageTarget(targetUnits: List<Unit>) =
        targetUnits.filter { it.alive }.shuffled().minByOrNull { it.power.value }

}
